sonixflasher.c

```diff
@@ -9,382 +9,788 @@
         sleep(3);
         attempt_no++;
     }
-    if(attempt_no > MAX_ATTEMPTS) return false;
-
-    clear_buffer(buf, 65);
-    read_bytes = hid_get_feature(dev, buf);
-    if(read_bytes != RESPONSE_LEN + 1)
-    {
-        fprintf(stderr, "ERROR: Failed to initialize: got response of length %d, expected %d.\n", read_bytes, RESPONSE_LEN);
-        return false;
-    }
-    bool reboot_fail = !read_response_32(buf, 0, &resp);
-    bool init_fail = !read_response_32(buf, CMD_INIT, &resp);
-    if(init_fail)
-    {
-        if(oem_reboot && reboot_fail)
-        {
+    if (attempt_no > MAX_ATTEMPTS) return false;
+
+    if (!hid_get_feature(dev, buf, CMD_GET_FW_VERSION)) return false;
+    chip = sn32_decode_chip(buf);
+    if (chip == 0) return false;
+    cs_level = sn32_get_cs_level(buf);
+    if (!sn32_check_isp_code_option(buf)) return false;
+
+    bool reboot_fail = !read_response_32(buf, 0, 0, &resp);
+    bool init_fail   = !read_response_32(buf, 0, CMD_VERIFY(CMD_GET_FW_VERSION), &resp);
+    if (init_fail) {
+        if (oem_reboot && reboot_fail) {
             fprintf(stderr, "ERROR: Failed to initialize: response cmd is 0x%08x, expected 0x%08x.\n", resp, 0);
-        }
-        else fprintf(stderr, "ERROR: Failed to initialize: response cmd is 0x%08x, expected 0x%08x.\n", resp, CMD_INIT);
+        } else
+            fprintf(stderr, "ERROR: Failed to initialize: response cmd is 0x%08x, expected 0x%08x.\n", resp, CMD_VERIFY(CMD_GET_FW_VERSION));
         return false;
     }
-    // // 2) Prepare for flash
+    return true;
+}
 
-    printf("Preparing for flash...\n");
+bool protocol_code_option_check(hid_device *dev) {
+    unsigned char buf[REPORT_SIZE];
+    // 02) Prepare for Code Option Table check
+    printf("\n");
+    printf("Checking Code Option Table...\n");
+    clear_buffer(buf, REPORT_SIZE);
+    buf[0] = CMD_COMPARE_CODE_OPTION;
+    write_buffer_16(buf + 1, CMD_BASE);
+    write_buffer_16(buf + 4, code_option);
+    if (!hid_set_feature(dev, buf, REPORT_SIZE)) return false;
+    clear_buffer(buf, REPORT_SIZE);
+    return true;
+}
 
-    clear_buffer(buf, 65);
-    write_buffer_32(buf, CMD_PREPARE);
-    write_buffer_32(buf+5, (uint32_t)offset);
-    write_buffer_32(buf+9, (uint32_t)(fw_size/64));
-    if(!hid_set_feature(dev, buf, 65)) return false;
+bool protocol_reset_cs(hid_device *dev) {
+    unsigned char buf[REPORT_SIZE];
+    // 03) Reset Code Option Table
+    printf("\n");
+    printf("Resetting Code Option Table...\n");
+    printf("Setting Code Option Table 0x%04x with CS 0x%04X.\n", code_option, CS0);
+    clear_buffer(buf, REPORT_SIZE);
+    buf[0] = CMD_SET_ENCRYPTION_ALGO;
+    write_buffer_16(buf + 1, CMD_BASE);
+    write_buffer_16(buf + 4, code_option);
+    write_buffer_16(buf + 6, CS0); // WARNING THIS SETS CS0
+    if (!hid_set_feature(dev, buf, REPORT_SIZE)) return false;
+    if (!hid_get_feature(dev, buf, CMD_SET_ENCRYPTION_ALGO)) return false;
+    clear_buffer(buf, REPORT_SIZE);
```

Comment:
Useless clearing of local stack buffer which is about to go out of scope. Same thing in next few functions.

Reply:
Mostly being extra careful with buffer management here while implementing undocumented bits of the protocol